deflate.c

```diff
@@ -160,12 +181,27 @@ struct static_tree_desc_s {int dummy;}; /* for buggy compilers */
 
 /* ===========================================================================
  * Update a hash value with the given input byte
- * IN  assertion: all calls to to UPDATE_HASH are made with consecutive
+ * IN  assertion: all calls to to hash are made with consecutive
  *    input characters, so that a running hash key can be computed from the
  *    previous key instead of complete recalculation each time.
  */
-#define UPDATE_HASH(s,h,c) (h = (((h)<<s->hash_shift) ^ (c)) & s->hash_mask)
+local uInt hash(s, h, str)
+    const deflate_state* s;
+    const uInt h;
+    const voidpf str;
+{
 
+#ifdef USE_CRC32_INTRINSIC
+#ifdef ARM_CRC32_INTRINSIC
+    (void)h;
+    return __crc32w(0, *(const uInt*)(str - (MIN_MATCH - 1))) & s->w_mask;
```

Comment:
Size of variable pointed by void * is undefined, so result of - operator is unpredictable... Should always use const char *, which gives the expected result. Also I think the function should be at least inlined because not all compilers inline short and/or simple functions by default and this can make it slower on some non-ARM targets.